### ISSUE TITLE
journalctl --follow does not exit immediately after stdout is closed

### ISSUE THREAD

**Issue by ibukanov** [systemd/systemd#9374]
**systemd version the issue has been seen with**
> systemd 238

**Used distribution**
> Fedora 28

**Expected behaviour you didn't see**
 > `journalctl -f` should exit immediately when its stdout was closed  

**Unexpected behaviour you saw**
 > After stdout was closed `journalctl -f` waits until it needs to write an output and only exits then.  
This makes problematic to use  `journalctl -f` in a pipeline together with grep or similar command that needs to do extra filtering and exit. In such case journalctl process may be running potentially forever.

**Steps to reproduce the problem**
Run the following shell script line:
```
(sleep 10; echo after_10s_sleep >&2; logger TEST_JOURNAL_EXIT) & (journalctl -f MESSAGE=TEST_JOURNAL_EXIT; echo journalctl_exit >&2) | sleep 1
```

The expected behaviour is that after 1s the script should print  `journalctl_exit`, then the shell should show a prompt to indicate that the pipeline ending with sleep terminated and then a message `after_10s_sleep` should appear from a pipeline running in the background.

The actual behaviour was that `after_10s_sleep` was printed first, then journalctl_exit and only then the shell shows the prompt.

**Comment by poettering** [comment]
Hmm, i guess we could hook this up, but i wonder what "tail -f" does in this case? journalctl -f is after all modelled after "tail -f", hence it probably makes sense to follow that...

**Comment by kirbyfan64** [comment]
[`tail` seems to do a `select` on stdout](https://github.com/coreutils/coreutils/blob/master/src/tail.c#L333):

```c
/* If the output has gone away, then terminate
as we would if we had written to this output. */
static void
check_output_alive (void)
{
  if (! monitor_output)
    return;

  struct timeval delay;
  delay.tv_sec = delay.tv_usec = 0;

  fd_set rfd;
  FD_ZERO (&rfd);
  FD_SET (STDOUT_FILENO, &rfd);

  /* readable event on STDOUT is equivalent to POLLERR,
and implies an error condition on output like broken pipe. */
  if (select (STDOUT_FILENO + 1, &rfd, NULL, NULL, &delay) == 1)
    raise (SIGPIPE);
}
```

**Comment by poettering** [comment]
ok, thanks for checking. We should definitely implement a similar logic then!

**Comment by ibukanov** [comment]
Interesting that tail -f from busybox shows a similar misfeature. Which is not surprising since their source https://github.com/mirror/busybox/blob/master/coreutils/tail.c shows no code to detect early closing of stdout. More thumbs up for thoughtfulness of GNU coreutils.

**Comment by poettering** [comment]
fix waiting in #10511
